fix(errdefs): support Wrap() []errors

stevenh opened the pull request:
Add support for `Wrap() []errors` as provided by `errors.Join` so that all `IsXXX` methods work correctly when provided with an error returned by `errors.Join`.

Also clarified that the search stops at the first match to a expected type in contrast to stdlib `errors.Is` which is exhaustive.

Fixes #48211

stevenh committed:
fix(errdefs): support Wrap() []errors

Add support for Wrap() []errors as provided by errors.Join so that all
IsXXX methods work correctly when provided with an error returned by
errors.Join.

Also clarified that the search stops at the first match to a expected type
in contrast to stdlib errors.Is which is exhaustive.

Fixes #48211

Signed-off-by: Steven Hartland <stevenmhartland@gmail.com>

stevenh commented:
All the test failing in CI are passing when run locally, if someone could point me in the right direction as to how to reproduce locally that would be appreciated.

For example CI failure:
```shell
https://github.com/moby/moby/actions/runs/10064397353/job/27822088022?pr=48220#step:15:850
=== FAIL: github.com/docker/docker/integration/container TestCreateVolumesFromNonExistingContainer (0.23s)
    create_test.go:521: assertion failed: expression is false: errdefs.IsInvalidParameter(err)
```

Local:
```
Running tool: /usr/local/go/bin/go test -timeout 10m -run ^TestCreateVolumesFromNonExistingContainer$ github.com/docker/docker/integration/container -count=1 -v

INFO: Testing against a local daemon
=== RUN   TestCreateVolumesFromNonExistingContainer
--- PASS: TestCreateVolumesFromNonExistingContainer (0.17s)
PASS
ok  	github.com/docker/docker/integration/container	0.274s
```

stevenh commented:
While this is done I'm wondering how valid it is given the functions don't behave how most people would expect, prioritising the first error instead of an exhaustive search. Thoughts?

thaJeztah commented on errdefs/is.go:
```diff
@@ -34,48 +38,76 @@ func getImplementer(err error) error {
 	case wrapErr:
 		return getImplementer(e.Unwrap())
+	case wrapErrs:
+		for _, err := range e.Unwrap() {
+			switch err := getImplementer(err).(type) {
```
I wonder if we should consider changing `getImplementer()` to return `bool, error`; the `bool` would be an indication "we matched a known type"; in that case we couldn't have to match interfaces again, and could return.

```go
for _, err := range e.Unwrap() {
    if ok, err :=  getImplementer(err) {
        return err
    }
}
```

Just thinking out loud though; I haven't benchmarked though to check if that would be more performant than the `.(type)` here!

I should also add that @dmcgowan and @jsternberg are working on consolidating some of the `errdefs` packages (moby, containerd, buildkit) to make them interoperable. I don't think that has to block this PR (work is still in progress).

As part of that work, I noticed this PR from @jsternberg that looked like and interesting approach as well (but more of an FYI, as making similar changes would make a much larger diff); https://github.com/containerd/errdefs/pull/12

stevenh replied:
I'll have a look, see if it improves things.

stevenh replied:
Seems like pretty good improvements:
```shell
benchstat before.log after.log 
goos: linux
goarch: amd64
pkg: github.com/docker/docker/errdefs
cpu: Intel(R) Core(TM) i9-14900HX
                                       │  before.log  │             after.log              │
                                       │    sec/op    │   sec/op     vs base               │
IsNotFound/nil-32                        2.720n ±  2%   1.502n ± 2%  -44.78% (p=0.002 n=6)
IsNotFound/multi-wrapped-not-found-32    193.7n ±  2%   145.2n ± 1%  -24.99% (p=0.002 n=6)
IsNotFound/cause-not-found-32            84.11n ±  3%   76.17n ± 3%   -9.45% (p=0.002 n=6)
IsNotFound/join-cause-not-found-32       265.0n ±  1%   216.9n ± 6%  -18.12% (p=0.002 n=6)
IsNotFound/join-cause-other-32           376.8n ±  3%   268.2n ± 1%  -28.81% (p=0.002 n=6)
IsNotFound/direct-other-32               73.81n ±  3%   70.31n ± 3%   -4.73% (p=0.004 n=6)
IsNotFound/wrapped-other-32              140.1n ±  1%   165.5n ± 3%  +18.09% (p=0.002 n=6)
IsNotFound/join-not-found-32             202.7n ±  1%   146.2n ± 2%  -27.90% (p=0.002 n=6)
IsNotFound/join-invalid-param-32         208.4n ±  7%   149.3n ± 1%  -28.36% (p=0.002 n=6)
IsNotFound/join-other-32                 307.8n ± 10%   256.8n ± 1%  -16.57% (p=0.002 n=6)
IsNotFound/join-cause-invalid-param-32   266.7n ±  6%   231.3n ± 5%  -13.25% (p=0.002 n=6)
IsNotFound/direct-not-found-32           8.204n ±  2%   9.855n ± 2%  +20.13% (p=0.002 n=6)
IsNotFound/wrapped-not-found-32          76.07n ±  6%   80.00n ± 4%        ~ (p=0.093 n=6)
IsNotFound/multi-wrapped-other-32        202.5n ±  6%   202.3n ± 1%        ~ (p=1.000 n=6)
geomean                                  105.1n         90.11n       -14.26%
```

stevenh replied:
> As part of that work, I noticed this PR from @jsternberg that looked like and interesting approach as well (but more of an FYI, as making similar changes would make a much larger diff); [containerd/errdefs#12](https://github.com/containerd/errdefs/pull/12)

I believe that is an incompatible change, as remove that first match approach as it's only looking for the single type not the set of types.

stevenh replied:
I did test with an updated version using generics and a 3 value return but it was slower.
```go
// state represents state of isInterface check.
type state byte

const (
	// match indicates a match for the type was found.
	match state = iota
	// noMatch indicates another expected type was found instead.
	noMatch
	// notFound indicates no expected was found.
	notFound
)

// isInterface searches err depth first for an error which matches the
// interface type T or one of the other expected types.
// It returns:
//   - match if T was the first expected type encountered.
//   - noMatch if another expected type was found before T.
//   - notFound if no expected type was found.
func isInterface[T any](err error) state {
	switch e := err.(type) {
	case T:
		return match
	case
		ErrNotFound,
		ErrInvalidParameter,
		ErrConflict,
		ErrUnauthorized,
		ErrUnavailable,
		ErrForbidden,
		ErrSystem,
		ErrNotModified,
		ErrNotImplemented,
		ErrCancelled,
		ErrDeadline,
		ErrDataLoss,
		ErrUnknown:
		return noMatch
	case causer:
		return isInterface[T](e.Cause())
	case wrapErr:
		return isInterface[T](e.Unwrap())
	case wrapErrs:
		for _, err := range e.Unwrap() {
			v := isInterface[T](err)
			if v == notFound {
				continue
			}

			return v
		}
		return notFound
	default:
		return notFound
	}
}

// IsNotFound returns true if the first instance of one of the expected types
// implements ErrNotFound, otherwise false.
func IsNotFound(err error) bool {
	return isInterface[ErrNotFound](err) == match
}
```
```shell
benchstat after.log generics.log 
goos: linux
goarch: amd64
pkg: github.com/docker/docker/errdefs
cpu: Intel(R) Core(TM) i9-14900HX
                                       │  after.log  │            generics.log            │
                                       │   sec/op    │   sec/op     vs base               │
IsNotFound/join-other-32                 256.8n ± 1%   232.4n ± 6%   -9.50% (p=0.002 n=6)
IsNotFound/direct-not-found-32           9.855n ± 2%   9.106n ± 4%   -7.60% (p=0.002 n=6)
IsNotFound/wrapped-other-32              165.5n ± 3%   145.4n ± 5%  -12.15% (p=0.002 n=6)
IsNotFound/join-invalid-param-32         149.3n ± 1%   161.1n ± 1%   +7.90% (p=0.002 n=6)
IsNotFound/cause-not-found-32            76.17n ± 3%   81.20n ± 4%   +6.62% (p=0.002 n=6)
IsNotFound/join-cause-not-found-32       216.9n ± 6%   242.0n ± 5%  +11.52% (p=0.002 n=6)
IsNotFound/join-cause-invalid-param-32   231.3n ± 5%   228.1n ± 3%        ~ (p=0.818 n=6)
IsNotFound/wrapped-not-found-32          80.00n ± 4%   86.28n ± 6%   +7.85% (p=0.004 n=6)
IsNotFound/multi-wrapped-other-32        202.3n ± 1%   213.5n ± 2%   +5.53% (p=0.002 n=6)
IsNotFound/multi-wrapped-not-found-32    145.2n ± 1%   196.4n ± 4%  +35.22% (p=0.002 n=6)
IsNotFound/join-not-found-32             146.2n ± 2%   166.8n ± 2%  +14.16% (p=0.002 n=6)
IsNotFound/join-cause-other-32           268.2n ± 1%   286.0n ± 2%   +6.64% (p=0.002 n=6)
IsNotFound/nil-32                        1.502n ± 2%   1.766n ± 5%  +17.54% (p=0.002 n=6)
IsNotFound/direct-other-32               70.31n ± 3%   89.93n ± 3%  +27.90% (p=0.002 n=6)
geomean                                  90.11n        96.52n        +7.11%
```

thaJeztah commented on errdefs/http_helpers.go:
```diff
@@ -9,6 +9,7 @@ func FromStatusCode(err error, statusCode int) error {
 	if err == nil {
 		return nil
 	}
+
```
Oh! Looks like a left-over change from a previous iteration; can you revert this change?

thaJeztah commented:
> While this is done I'm wondering how valid it is given the functions don't behave how most people would expect, prioritising the first error instead of an exhaustive search. Thoughts?

That's intentional; errdefs can originate from deeper down in the stack, and have a different meaning in different subsystems.

Consider this scenario;

```
docker run --network=container:<other container> ...
```

In this example, the user wants to run a container, but using the networking namespaces of `<other container>`.

Now, assume this;

```bash
docker run --network=container:non-existing-container ...
```

The specified container (`non-existing-container`) doesn't exist, so when we try to check if it does, we get a `errdefs.ErrNotFound`; that's the correct error when looking up a non-existing container, but it's _NOT_ the correct type of error for the `docker run --network=container:...`; that action should not return a `NotFound` error, but (likely) an `ErrInvalidParameter`; the passed argument (`network:non-existing-container`) is invalid, because the specified container doesn't exist.

So in that case we wrap the original `ErrNotFound` error into a `ErrInvalidParameter`. The `ErrNotFound` error is no longer (directly) relevant (but still available in the stack).

stevenh commented:
> > While this is done I'm wondering how valid it is given the functions don't behave how most people would expect, prioritising the first error instead of an exhaustive search. Thoughts?
> 
> That's intentional; errdefs can originate from deeper down in the stack, and have a different meaning in different subsystems.
...
> 
> So in that case we wrap the original `ErrNotFound` error into a `ErrInvalidParameter`. The `ErrNotFound` error is no longer (directly) relevant (but still available in the stack).

I suspected that was the case, thanks for the confirmation, do my changes to doc comments make that intent clearer?

stevenh committed:
perf(errdef): optimise return check

Optimise the return check to avoid type cast.

benchstat before.log after.log
goos: linux
goarch: amd64
pkg: github.com/docker/docker/errdefs
cpu: Intel(R) Core(TM) i9-14900HX
                                       │  before.log  │             after.log              │
                                       │    sec/op    │   sec/op     vs base               │
IsNotFound/nil-32                        2.720n ±  2%   1.502n ± 2%  -44.78% (p=0.002 n=6)
IsNotFound/multi-wrapped-not-found-32    193.7n ±  2%   145.2n ± 1%  -24.99% (p=0.002 n=6)
IsNotFound/cause-not-found-32            84.11n ±  3%   76.17n ± 3%   -9.45% (p=0.002 n=6)
IsNotFound/join-cause-not-found-32       265.0n ±  1%   216.9n ± 6%  -18.12% (p=0.002 n=6)
IsNotFound/join-cause-other-32           376.8n ±  3%   268.2n ± 1%  -28.81% (p=0.002 n=6)
IsNotFound/direct-other-32               73.81n ±  3%   70.31n ± 3%   -4.73% (p=0.004 n=6)
IsNotFound/wrapped-other-32              140.1n ±  1%   165.5n ± 3%  +18.09% (p=0.002 n=6)
IsNotFound/join-not-found-32             202.7n ±  1%   146.2n ± 2%  -27.90% (p=0.002 n=6)
IsNotFound/join-invalid-param-32         208.4n ±  7%   149.3n ± 1%  -28.36% (p=0.002 n=6)
IsNotFound/join-other-32                 307.8n ± 10%   256.8n ± 1%  -16.57% (p=0.002 n=6)
IsNotFound/join-cause-invalid-param-32   266.7n ±  6%   231.3n ± 5%  -13.25% (p=0.002 n=6)
IsNotFound/direct-not-found-32           8.204n ±  2%   9.855n ± 2%  +20.13% (p=0.002 n=6)
IsNotFound/wrapped-not-found-32          76.07n ±  6%   80.00n ± 4%        ~ (p=0.093 n=6)
IsNotFound/multi-wrapped-other-32        202.5n ±  6%   202.3n ± 1%        ~ (p=1.000 n=6)
geomean                                  105.1n         90.11n       -14.26%

Signed-off-by: Steven Hartland <stevenmhartland@gmail.com>

stevenh commented:
CI error is a timeout which looks unrelated.
